src/bounces/test.rs

```diff
@@ -0,0 +1,189 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+use super::*;
+use auth_db::{Db, DbError};
+use settings::Settings;
+
+pub struct DbMockNoBounce;
+
+impl Db for DbMockNoBounce
+{
+  fn get_email_bounces(&self, _address: &str) -> Result<Vec<BounceRecord>, DbError>
+  {
+    let now = now_as_milliseconds();
+    Ok(vec![
+      BounceRecord {
+        bounce_type: BounceType::Hard,
+        created_at: now - 86400001,
+      },
+      BounceRecord {
+        bounce_type: BounceType::Soft,
+        created_at: now - 300001,
+      },
+      BounceRecord {
+        bounce_type: BounceType::Complaint,
+        created_at: now - 86400001,
+      },
+    ])
+  }
+}
+
+fn now_as_milliseconds() -> u64
+{
+  let now = SystemTime::now()
+    .duration_since(SystemTime::UNIX_EPOCH)
+    .expect("system time error");
+  now.as_secs() * 1000
+}
+
+#[test]
+fn check_no_bounces()
```

Comment:
The auth-server has some interesting test-cases for situations that shouldn't bounce, such as:

  * https://github.com/mozilla/fxa-auth-server/blob/9731a080587f24deaec1155b82d8679abd8a2bc4/test/local/bounces.js#L97
  * https://github.com/mozilla/fxa-auth-server/blob/9731a080587f24deaec1155b82d8679abd8a2bc4/test/local/bounces.js#L122

It's not clear to me whether these are equivalently covered by the tests above, but they're probably worth adding if not.

Reply:
Yeah, I knew my tests here were insufficient but was lazily avoiding doing the legwork that would make better tests possible.

I need to use the trait trick on `Settings`, so that tests can pass in custom config without relying on the real config that happens to be on disk. Overriding the array items in that part of the config via environment variables works, but it assumes deep knowledge of the resolved config that will fail if `dev.json` or `local.json` override stuff.

This will be useful for testing other stuff too though, so it's good that I finally get round to it!

Reply:
> I need to use the trait trick on `Settings`...

Fwiw it was much easier than that, I just needed to derive `Default` and then assign directly to `bouncelimits` on the result of `Settings::default`.